batch/deployment.yaml

```diff
@@ -42,9 +42,15 @@ spec:
          - -m
          - batch.driver
         resources:
-          requests:
-            memory: "1G"
-            cpu: "1"
+{% if deploy %}
+           requests:
+             memory: 375Mi
+             cpu: 100m
+{% else %}
+           requests:
+             memory: 37Mi
+             cpu: 10m
```

Comment:
Should you add limit here?  Can we boost the memory?  batch-driver is currently running at 86Mi and it will definitely need more memory when there are many instances, since the instance data structures are stored in memory.

Also, batch-driver is bursty and definitely uses more than 10m when it is running.  Again, my question about requests stands.  If batch-driver gets packed into a 10mCPU slot because of this and it is trying to schedule a massive batch, will it peg at 10m and nothing happens?  batch-driver isn't horizontally scaled.  That seems bad.  It seems like batch-driver request also limits or scheduling rate.

Reply:
No, that's not how requests work. The scheduler only sees requests, so batch-driver might get packed in with lots of other pods. batch-driver is allowed to exceed its request (eating into the requests of other pods) *as long as the node has resources to spare*. If, however, the node has no spare resources (say, it is full and everyone is using 100% of its request, or it is not full, but a few people are exceeding their requests significantly), then it will kill & reschedule pods that are exceeding their requests.

"peg" behavior is only created by a resource limit.

Now, what I'm not sure is how the autoscaler relates to this. For this system to work, we need the auto-scaler to select the number of nodes based on *actual* use not *request* or *limit*.

I'm inclined to take this aggressive step and then adjust based on our experience. I think it's clear that even after these changes we'll have *lots* of unused CPU (because most pods will use below their request).

Reply:
I'll add a limit though.